File: go/q-86-partition-list/q-86-partition-list.go
package main

import "fmt"

//86.分隔链表.partition-list

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
	p, q := &ListNode{}, &ListNode{}
	l, r := p, q
	for head != nil {
		if head.Val < x {
			p.Next = head
			p = p.Next
			head = head.Next
			p.Next = nil
		} else {
			q.Next = head
			q = q.Next
			head = head.Next
			q.Next = nil
		}
	}
	p.Next = r.Next
	return l.Next
}

//leetcode submit region end(Prohibit modification and deletion)

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	var x int = 3
	head := &ListNode{
		Val:  1,
		Next: nil,
	}
	result := partition(head, x)
	fmt.Println(result)
}

File: cpp/q-86-partition-list/q-86-partition-list.cpp
#include "iostream"
#include "vector"
using namespace std;

//86.分隔链表.partition-list

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode *p = new ListNode(), *q = new ListNode();
        ListNode *l = p, *r = q;
        while (head){
            ListNode *t = head;
            if (head->val < x){
                p->next = head;
                p = p->next;
            } else{
                q->next = head;
                q = q->next;
            }
            head = head->next;
            t->next = nullptr;
        }
        p->next = r->next;
        return l->next;
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(){
    vector<int> v = {1, 4, 3, 2, 5, 2};
    ListNode list, *p = &list;

    for (int i = 0; i < v.size(); ++i,p = p->next) {
        p->next = new ListNode(v[i], nullptr);
    }
    cout << (new Solution())->partition(list.next, 3) << endl;
    return 0;
}
